lib/view_component/css_module.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+require "sass"
+
+module ViewComponent
+  class CSSModule < Sass::Tree::Visitors::Base
+    def initialize(module_name, raw_css)
+      @module_name, @mappings = module_name, {}
+
+      @css_root = Sass::SCSS::CssParser.new(raw_css, "(CSSModules)", 1).parse
+
+      @module_hash = compute_hash
+    end
+
+    # Given a `module_name` and `raw_css`, rewrite `raw_css`
+    # apply opaque transformations to `raw_css` so that
+    # selectors can only be accessed programatically,
+    # not by class name literals.
+    def self.rewrite(*args)
+      new(*args).rewrite
+    end
+
+    def rewrite
+      Sass::Tree::Visitors::SetOptions.visit(@css_root, {})
+
+      {
+        mappings: mappings,
+        css: @css_root.render
+      }
+    end
+
+    private
+
+    # Generate a short, random-ish token to prevent CSS selector collisions.
+    def compute_hash
+      Digest::MD5.hexdigest(@module_name).first(5) # rubocop:disable GitHub/InsecureHashAlgorithm
```

Comment:
`MD5` is pretty good at generating (mostly) collision free hashes, but a couple of points:

 - It's counted as encryption and federal compliance regulations have specific rubrics around encryption standards. Using this for "enterprise" software involves some bureaucratic complexity. It might be easier to migrate from this just to avoid having the conversation.
 - JS does not have MD5 in its standard library. It's unlikely to get it as it's now an obsolete standard. It's also a non-trivial implementation; it requires quite a lot of code to implement or pull in, so if we need to mirror these on the client side we'd be looking at adding 1-2kb of JS code just to do this.

Perhaps a better option is to use a simpler digest mechanism, such as `base64`? is probably more ubiquitous than MD5, and where it isn't it can be implemented in a dozen or so LOC. It does have a much higher potential for collisions; but given it's used as _part_ of a class selector here, the risk of collisions is likely the same as using MD5.

Reply:
The problem with simple truncated `base64` is that you'll have the same value if components have the same suffix -- it's only encoding the string with a different radix, not hashing it:

```rb
Base64.encode64("HelloWorld").first(5)
=> "SGVsb"
Base64.encode64("HelloThere").first(5)
=> "SGVsb"
```

I'm not sure if there is some other hashing function that does not carry the same compliance regulations you mention?

Reply:
I don't think this is necessarily a problem as the class names would still be different as `@module-name` is used both undigested and digested parts of the selector:  a full class would be `.HelloThere_SGVsb_foo` or `.HelloWorld_SGVsb_foo`. The intent of the `compute_hash` function seems to be to provide enough entropy in a class name so as not to collide with human authored classes; `.HelloThere_foo` is more likely to already exist in existing CSS.